flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/type/FlinkCalciteTableMappingRule.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.calcite.sql.type;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/** Rules that determine whether a type is castable from another type. */
+public class FlinkCalciteTableMappingRule implements SqlTypeMappingRule {
+    private static final FlinkCalciteTableMappingRule INSTANCE;
```

Comment:
The behavior of `SqlTypeCoercionRule` was changed (https://issues.apache.org/jira/browse/CALCITE-5662) and started leading to a number of failures in `CastFunctionITCase`.
New mapping rule makes it working in same way as before